[IMP] stock: merge product name and move description in one column

mohammadhossam opened the pull request:
This commit applies the new product name and description widget introduced in 216634f on the stock moves of a stock picking. The widget combines the name of the product and the description of the line in one cell for better readability and saving space. Since this widget was previously only used in invoices, sale orders and purchase orders, it was set up to be used with a record whose lines could be sections or notes (which is not the case with stock pickings). Therefore, the code
of the widget was refactored such that the common functionality is moved to the `product` module and only the customizations related to (not) having sections and notes feature are left in `account` and `stock` modules.

odoo/enterprise#68629

Task-4126824

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/177390.png)](https://mergebot.odoo.com/odoo/odoo/pull/177390)

svs-odoo commented on addons/stock/models/stock_picking.py:
```diff
@@ -1074,10 +1074,10 @@ def _onchange_picking_type(self):
-                if not move.product_id:
-                    continue
-                move.description_picking = move.product_id._get_description(move.picking_type_id)
+            # for move in (self.move_ids | self.move_ids_without_package):
```
You comment those lines in the first picking and uncomment them in the second picking.
Don't comment them at all in the first picking instead please

svs-odoo replied:
@mohammadhossam Please don't mark as resolved a comment if you didn't do the change.
PS: I edited my previous comment, I wanted to mean "the first commit", not "the first picking" :grimacing:

mohammadhossam replied:
Oh, I didn't notice that the change was split across both commits. I removed the change completely :+1:

svs-odoo reviewed (commented):
Yop ! First review :grin: 
I noticed when hide description for the PO lines, the description is actually hidden.
But for the picking moves, hide the description has no effect :/

About the first commit, its name is no good, it should refer to the modified modules, refer to `stock_barcode` in the header in community makes no sense regarding this rule

mohammadhossam commented on addons/purchase/models/purchase_order_line.py:
```diff
@@ -73,6 +73,7 @@ class PurchaseOrderLine(models.Model):
         ('line_section', "Section"),
         ('line_note', "Note")], default=False, help="Technical field for UX purpose.")
     is_downpayment = fields.Boolean()
+    selected_seller_id = fields.Many2one('product.supplierinfo', compute='_compute_selected_seller_id', help='Technical field to store the vendor pricelist used to generate this line')
```
I though it could be a good idea to have it here so that it's usable in the future in case we need it in another flow. I know it can be replaced with a simple call to _select_seller but I thought also it's good to have it as a nice wrapper instead of calling _select_seller every time we need it, specially that it takes a lot of parameters. What do you think?

svs-odoo replied:
After a second though, I think the field itself is a good idea, and it could stay in `purchase` if we clean the code and use this field where it can be used, instead of calling `_select_seller`.
But for the sake of the merging of this PR, it could be better to stick to `_select_seller` now and create this field and do the clean in another PR.
(If you decide to do it in this PR anyway, create the field and do the clean in another commit please, so we can easily separate the commit for the task and the commit for the clean)

mohammadhossam replied:
Okay, I moved it now into a separate commit.

svs-odoo commented on addons/stock/models/stock_rule.py:
```diff
@@ -323,10 +325,12 @@ def _get_stock_move_values(self, product_id, product_qty, product_uom, location_
-        product_id = product_id.with_context(lang=(partner and partner.lang) or self.env.user.lang)
-        picking_description = product_id._get_description(self.picking_type_id)
+        if partner:
+            product_id = product_id.with_context(lang=(partner and partner.lang) or self.env.user.lang)
```
Same remark than [here](https://github.com/odoo/odoo/pull/177390#discussion_r2100725520), if you undo a change, undo it in the commit you make the change in the first place, not in the second commit.
Even if at the end it seems there is no change, actually if the commit are taken separately, the change happens and so you rewrite the blame history on those line for nothing

mohammadhossam replied:
Done :+1:

Feyensv commented on addons/sale/models/sale_order_line.py:
```diff
@@ -427,7 +427,7 @@ def _get_sale_order_line_multiline_description_variants(self):
             return ""
 
-        name = "\n"
+        name = ""
```
Why do you need to change this line here? :thinking:

svs-odoo commented:
I tested again and found two issues:
1. In SO, every time you update a SO line description, the product's name is added to it at the save:
![Screenshot from 2025-06-13 12-53-00](https://github.com/user-attachments/assets/02024fc6-ea39-4a10-a9fd-ae456d0cafbd)

2. When confirming a PO, the description on receipt move lines contains some `[False] False`:
![Screenshot from 2025-06-13 12-46-46](https://github.com/user-attachments/assets/fd82486d-ee90-45ce-b2a0-355a677737cc)
Accessory, some of the description's lines are kind of duplicate:
![Screenshot from 2025-06-13 12-47-36](https://github.com/user-attachments/assets/69aad475-f85e-4f72-b677-8ced14c68e07)
The `(Black)` part is already on the product's name.
That's why we need to have a python test for this part, to ensure the description is always what we expect.

mohammadhossam commented:
Hi @svs-odoo 
I fixed both issues that you mentioned. I also removed duplicate info from the description. I added tests for both purchase and sale scenarios.

svs-odoo commented on addons/purchase_stock/tests/test_purchase_lead_time.py:
```diff
@@ -197,8 +198,8 @@ def test_merge_po_line_3(self):
         purchase_order.button_confirm()
-        self.assertEqual(purchase_order.picking_ids[0].move_ids_without_package.filtered(lambda x: x.product_uom_qty == 15).description_picking, t_shirt.display_name + "\n" + order_line_description + "Color (Red)", 'wrong description in picking')
-        self.assertEqual(purchase_order.picking_ids[0].move_ids_without_package.filtered(lambda x: x.product_uom_qty == 10).description_picking, t_shirt.display_name + "\n" + order_line_description + "Color (Green)", 'wrong description in picking')
+        self.assertEqual(purchase_order.picking_ids[0].move_ids_without_package.filtered(lambda x: x.product_uom_qty == 15).description_picking, t_shirt.display_name + "\n" + order_line_description, 'wrong description in picking')
+        self.assertEqual(purchase_order.picking_ids[0].move_ids_without_package.filtered(lambda x: x.product_uom_qty == 10).description_picking, t_shirt.display_name + "\n" + order_line_description, 'wrong description in picking')
```
You can remove definition of `order_line_description` (line 175) and write the description in clear.
The definition of the `order_line_description` variable depends of a condition which is a bad practice in a test and its name is confusing (order line description... Purchase order ? Nope, it's the receipt description)
```suggestion
        self.assertEqual(purchase_order.picking_ids[0].move_ids_without_package.filtered(lambda x: x.product_uom_qty == 15).description_picking, t_shirt.display_name + "\n" + 'Receive with care', 'wrong description in picking')
        self.assertEqual(purchase_order.picking_ids[0].move_ids_without_package.filtered(lambda x: x.product_uom_qty == 10).description_picking, t_shirt.display_name + "\n" + 'Receive with care', 'wrong description in picking')
```

svs-odoo commented on addons/sale_stock/tests/test_sale_stock.py:
```diff
@@ -21,7 +21,7 @@ def setUpClass(cls):
         })
 
-    def _get_new_sale_order(self, amount=10.0, product=False):
+    def _get_new_sale_order(self, amount=10.0, product=False, sol_vals={}):
```
We should avoid to use mutable as default parameter's value ([reading about this issue](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python))
```suggestion
    def _get_new_sale_order(self, amount=10.0, product=False, sol_vals=False):
```
And then, you can do that instead:
```python
sol_vals = sol_vals or {}
```

svs-odoo commented on addons/sale_stock/tests/test_sale_stock.py:
```diff
@@ -32,10 +32,11 @@ def _get_new_sale_order(self, amount=10.0, product=False):
                 'product_id': product.id,
                 'product_uom_qty': amount,
-                'price_unit': product.list_price})
+                'price_unit': product.list_price,
```
You can take the opportunity to update create line to use the command instead :)
```python
            'order_line': [Command.create({
```

svs-odoo commented on addons/stock/static/src/views/picking_form/stock_move_one2many.js:
```diff
@@ -0,0 +1,76 @@
+        this.descriptionColumn = "description_picking";
+    }
+
+    get completeLabel() {
+        return this.props.record.data[this.descriptionColumn];
+    }
+
```
Seems unused
```suggestion
```

svs-odoo commented on addons/stock/static/src/views/picking_form/stock_move_one2many.js:
```diff
@@ -0,0 +1,76 @@
+          [this.descriptionColumn]: value,
+        });
+    }
+
+    get doneOrCancelled() {
+        return ["done", "cancel"].includes(this.props.record.data.state);
+    }
```
Seems to be unused too
```suggestion
```

svs-odoo commented on addons/stock/static/src/views/picking_form/stock_move_one2many.js:
```diff
@@ -0,0 +1,76 @@
+        return label.trim();
+    }
+
+    get nonDraftTextareaVisible() {
+        return (this.columnIsProductAndLabel.value && this.label)
+            || (!this.columnIsProductAndLabel.value && !this.productName && this.label)
+            || this.labelVisibility.value;
+    }
+
```
Same
```suggestion
```

svs-odoo commented on addons/stock/static/src/views/picking_form/stock_move_one2many.js:
```diff
@@ -0,0 +1,76 @@
+
+registry.category("fields").add("stock_move_one2many", stockMoveX2ManyField);
+
+export class MoveProductLabelField extends ProductNameAndDescriptionField {
```
Separate the file in two: one for `stock_move_one2many` widget and one for `move_product_label_field` widget

robodoo commented:
@mohammadhossam @svs-odoo because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

svs-odoo commented:
robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
Merge method set to rebase and fast-forward.

Feyensv commented on addons/sale/static/tests/sale_product_field.test.js:
```diff
@@ -86,7 +86,7 @@ test("Hide product name if its not translated", async () => {
         order_line: [Command.create({
             product_id: product.id,
-            name: product.name.concat("\nA description"),
+            name: "A description",
```
Why this change?
What is effectively stored in the name is the concatenation of the name and the description, so I don't understand why you'd need to change this :thinking: 

cc: @lipiraux
